File: 2019W14/preppindataw14.R
library(dplyr)
library(tidyr)
library(readr)

orders <- read_csv("E:/Cafe Orders by Product.csv",
                  col_types = cols(TicketID = col_character(),
                                   MemberID = col_character())) %>%
  mutate_at(vars('MemberID'), ~replace_na(., '0')) %>%
  mutate_at(vars('Price'), ~replace_na(., 1.5)) 

final <- orders %>%
  group_by(`TicketID`) %>%
  mutate('Total Ticket Price' = sum(`Price`),
         'Count Type'= n_distinct(`Type`)) %>%
  filter(`Count Type`==3) %>%
  group_by(`TicketID`,`Type`) %>%
  summarise('Avg Type Price' = mean(`Price`),
            'Count Meal'= n(),
            'MemberID' = first(`MemberID`),
            'Total Ticket Price' = first(`Total Ticket Price`)) %>%
  group_by(`TicketID`) %>%
  mutate('Meal Deal Count' = min(`Count Meal`),
         'Excess Count' = `Count Meal` - min(`Count Meal`)) %>%
  summarise('MemberID' = first(`MemberID`),
            'Total Ticket Price' = first(`Total Ticket Price`),
            'Meal Deal Earnings' = 5*first(`Meal Deal Count`),
            'Total Excess' = sum(`Excess Count`*`Avg Type Price`)) %>%
  mutate('Tickets Price Variance to Meal Deal Earnings'=`Total Ticket Price`-`Meal Deal Earnings`-`Total Excess`) 

View(final)

File: 2019W14/PreppindataW14.py
import pandas as pd

orders = pd.read_csv("E:\Cafe Orders by Product.csv")
meal_deal_price = 5

orders[['Price']] = orders[['Price']].fillna(value=1.5)
orders[['MemberID']] = orders[['MemberID']].fillna(value='0')

ticket = orders.groupby(['TicketID'], as_index=False).agg({'Price':'sum','Type':'nunique'})
ticket.rename(columns={'Price':'Total Ticket Price'}, inplace=True)
ticket = ticket[ticket['Type']==3]
ticket = ticket[['TicketID','Total Ticket Price']]
orders= orders.merge(ticket,on='TicketID', how='inner')

check = orders.groupby(['TicketID','Type']).size().to_frame(name='Type Count')
check = (check
         .join(orders.groupby(['TicketID','Type']).agg({'Price':'mean'})).rename(columns={'Price': 'TypeAvgPrice'})
         .join(orders.groupby(['TicketID','Type']).agg({'MemberID':'first'}))
         .join(orders.groupby(['TicketID','Type']).agg({'Total Ticket Price':'first'}))
         .reset_index()
         )
check['Meal Deal Count'] = check['Type Count'].groupby(check['TicketID']).transform('min')
check['Meal Deal Earning'] = check['Meal Deal Count']*meal_deal_price
check['Excess Earning'] = (check['Type Count']-check['Meal Deal Count'])*check['TypeAvgPrice']

final = check.groupby(['TicketID'], as_index=False).agg({'MemberID':'first',
                     'Total Ticket Price':'first',
                     'Meal Deal Earning':'first',
                     'Excess Earning':'sum'})
final.rename(columns={'Meal Deal Earning':'Total Meal Deal Earnings',
                      'Excess Earning':'Total Excess'}, inplace=True)
final['Tickets Price Variance to Meal Deal Earnings'] = (
        final['Total Ticket Price'] - final['Total Meal Deal Earnings']-final['Total Excess'] 
        )
